:bug: Fix cache sync timeout functionality

alvaroaleman opened the pull request:
So far, timing out the cache sync in most realistic scenarios didn't work,
bceause source.Kind gets an already started cache from the Manager. A
cache that is already started will block forever on GetInformer which we
called in source.Kinds start and not its WaitForSync. The context passed
to Start however defines the entire lifecycle of the Source, not the
Start timeout.

This change makes source.Kind call GetInformer in Start but in a new
go routine and WaitForSync just wait for that to finish or for its
context to be cancelled. This preserves the existing semantic of
starting in Start but waint for ready in WaitForSync.

<!-- please add an icon to the title of this PR (see VERSIONING.md), and delete this line and similar ones -->
<!-- the icon will be either ⚠ (:warning:, major), ✨ (:sparkles, minor), 🐛 (:bug:, patch), 📖 (:book:, docs), or 🌱 (:seedling:, other) -->

<!-- What does this do, and why do we need it? -->
/cc @varshaprasad96 @estroz @vincepri 

Fixes https://github.com/kubernetes-sigs/controller-runtime/issues/1346

alvaroaleman commented on pkg/internal/controller/controller_test.go:
```diff
@@ -104,40 +106,21 @@ var _ = Describe("controller", func() {
 			close(done)
 		})
 
-		It("should wait for each informer to sync", func(done Done) {
```
The only thing this test ever tested was that `controller.Start` with a cancelled context doesn't error. With the changes in this PR however it will always error because the context passed from the Controller to WaitForSync is cancelled

estroz replied:
In that case do you think it's worth having a test case that checks that an error is always returned if the context is cancelled?

alvaroaleman replied:
IMHO the only thing someone who starts a Controller with a cancelled context can realistically expect is that it ends quickly, if it errors or not along the way ends up being an implementation detail. It is not going to result in anything useful.

alvaroaleman commented on pkg/source/source_test.go:
```diff
@@ -90,6 +90,7 @@ var _ = Describe("Source", func() {
 					},
 				}, q)
 				Expect(err).NotTo(HaveOccurred())
+				Expect(instance.WaitForSync(context.Background())).NotTo(HaveOccurred())
```
We need to call WaitForSync here because before that finished, the Informer won't have the EventHandler attached. Not waiting makes the test racy.

alvaroaleman commented:
apidiff fails because:
```
  Incompatible changes:
  - Kind: old is comparable, new is not
```

I don't think Kind being comparable is something we consider part of the API, WDYT?

estroz commented on pkg/source/source.go:
```diff
@@ -110,16 +115,29 @@ func (ks *Kind) Start(ctx context.Context, handler handler.EventHandler, queue w
+		// Lookup the Informer from the Cache and add an EventHandler which populates the Queue
+		i, err := ks.cache.GetInformer(ctx, ks.Type)
+		if err != nil {
+			if kindMatchErr, ok := err.(*meta.NoKindMatchError); ok {
```
```suggestion
			kindMatchErr := &meta.NoKindMatchError{}
			if errors.As(err, &kindMatchErr) {
```

alvaroaleman commented:
@estroz addressed your comment, ptal. I've also renamed the pull as the issue wasn't with the cache sync per se but with the cache sync timeout.

Just to mention that, another possible approach to this would have been to make the `GetInformer` call not wait for the given informer to be synced after the cache is started and only do that in Get/List, that would allow folks like ourselves to have one non-blocking and one explicitly blocking call. The reason I didn't do that is that it might break others that rely on GetInformer blocking. WDYT?

vincepri reviewed (approved):
/approve
/lgtm

vincepri commented:
/hold 

for @estroz

estroz commented:
>Just to mention that, another possible approach to this would have been to make the GetInformer call not wait for the given informer to be synced after the cache is started and only do that in Get/List, that would allow folks like ourselves to have one non-blocking and one explicitly blocking call.

@alvaroaleman that depends, can you think of any reason `GetInformer()` returning an unsync'd cache would be beneficial?

The current approach is fine with me.

/lgtm

alvaroaleman commented:
/hold cancel

alvaroaleman commented:
/retest

alvaroaleman commented:
/retest
